Breeze.TumbleBit.Client/TumbleBitManager.cs

```diff
@@ -336,46 +336,78 @@ public enum TumbleState
+                // Since the list is shuffled, we can simply iterate through it and try each server until one is valid & reachable.
                 foreach (RegistrationRecord record in registrations)
                 {
-                    this.TumblerAddress = "ctb://" + record.Record.OnionAddress + ".onion?h=" + record.Record.ConfigurationHash;
+                    this.TumblerAddress = $"ctb://{record.Record.OnionAddress}.onion?h={record.Record.ConfigurationHash}";
+
+                    //Do not attempt a connection to the Masternode which is blacklisted
+                    if (masternodeBlacklist != null && masternodeBlacklist.Contains(this.TumblerAddress)) {
+                        this.logger.LogDebug($"Skipping connection attempt to blacklisted masternode {this.TumblerAddress}");
+                        continue;
+                    }
 
-                    var attemptConnection = await TryUseServer();
+                    try
+                    {
+                        var tumblerParameterResult = await TryUseServer();
 
-                    if (!attemptConnection.Failure)
+                        if (tumblerParameterResult.Success)
+                        {
+                            return tumblerParameterResult;
+                        }
+                    }
+                    catch (Exception e)
                     {
-                        return attemptConnection;
+                        this.logger.LogDebug(e, $"Unable to connect to masternode: {this.TumblerAddress}");
                     }
                 }
 
-                // If we reach this point, no servers were reachable
-                this.logger.LogDebug("Did not find a valid registration");
+                // If we reach this point, no servers were reachable.
+                this.logger.LogDebug($"Attempted connection to {registrations.Count} masternodes and did not find a valid registration");
                 return Result.Fail<ClassicTumblerParameters>("Did not find a valid registration");
             }
             else
             {
-                return await TryUseServer();
+                try
+                {
+                    var tumblerParameterResult = await TryUseServer();
+
+                    if (tumblerParameterResult.Success)
+                    {
+                        return tumblerParameterResult;
+                    }
+                }
+                catch (Exception e)
+                {
+                    this.logger.LogDebug(e, $"Unable to connect to masternode: {this.TumblerAddress}");
+                }
+
+                // Blacklist masternode address which we have just failed to connect to so that
+                // we won't attempt to connect to it again in the next call to ConnectToTumblerAsync.
+                HashSet<String> blacklistedMasternodes = new HashSet<String>() { this.TumblerAddress };

```

Comment:
`blacklistedMasternodes` unused local variable.  Do you need to add this to the HashSet and persist this in a class level field ?

Reply:
also use string type keyword ... `HashSet<string>...`

Reply:
This is a defect, the blacklistedMasternodes should be passed as an argument to ConnectToTumblerAsync